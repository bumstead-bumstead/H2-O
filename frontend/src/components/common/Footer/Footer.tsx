import { type PropsWithChildren } from 'react';
import styled from '@emotion/styled';
import { CTAButton, Flex, PriceSummaryButton, SummaryPopup, Typography } from '@/components/common';
import { useSafeContext, useToggle } from '@/hooks';
import { toSeparatedNumberFormat } from '@/utils/number';
import { SelectionContext } from '@/providers/SelectionProvider';
import { SlideContext } from '@/providers/SlideProvider';

interface Props {
  isSticky?: boolean;
}
function Footer({ children, isSticky = false }: PropsWithChildren<Props>) {
  const { totalPrice } = useSafeContext(SelectionContext);
  const { setCurrentSlide } = useSafeContext(SlideContext);
  const { status: isPopupOpen, setOff: handleClosePopup, setOn: handleOpenPopup } = useToggle(false);

  const handleClickNext = () => {
    setCurrentSlide((prev) => prev + 1);
  };

  return (
    <FooterContainer
      data-testid='footer'
      as='footer'
      justifyContent='flex-end'
      height={120}
      width='100%'
      paddingTop={20}
      isSticky={isSticky}
    >
      <ContentContainer justifyContent={children ? 'space-between' : 'flex-end'}>
        {children}
        <Flex flexDirection='column' alignItems='flex-end' width='fit-content'>
          <Flex justifyContent='space-between' alignItems='center' width='100%' marginBottom={6} gap={20}>
            <PriceSummaryButton onClick={handleOpenPopup}>견적요약</PriceSummaryButton>
            <Flex alignItems='center' gap={8}>
              <Typography font='TextKRRegular12' color='gray700'>
                현재 총 가격
              </Typography>
              <Typography font='HeadKRMedium24' color='primary500'>{`${toSeparatedNumberFormat(
                totalPrice,
              )} 원`}</Typography>
            </Flex>
          </Flex>
          <CTAButton onClick={handleClickNext}>다음</CTAButton>
        </Flex>
      </ContentContainer>
      {isPopupOpen && <SummaryPopup handleClickCloseButton={handleClosePopup} />}
    </FooterContainer>
  );
}

export default Footer;

const FooterContainer = styled(Flex)<Pick<Props, 'isSticky'>>`
  position: ${({ isSticky }) => (isSticky ? 'sticky' : 'fixed')};
  bottom: 0;
  background-color: ${({ theme }) => theme.colors.footerBg};
`;

const ContentContainer = styled(Flex)`
  width: 1024px;
  margin: 0 auto;
`;
