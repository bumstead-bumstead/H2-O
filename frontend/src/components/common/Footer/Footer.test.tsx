import Footer from './Footer';
import { render, screen } from '@/tests/test-util';

// TODO: 버튼 팝업 오픈 테스트 추가해야함, 이상하게 click 했음에도 팝업이 띄워지지 않아 일단 보류
describe('Footer 컴포넌트 테스트', () => {
  beforeAll(() => {
    const portalRoot = document.createElement('div');
    portalRoot.setAttribute('id', 'portal-root');
    document.body.appendChild(portalRoot);
  });

  it('Footer 정상적으로 렌더링된다.', () => {
    render(<Footer />);

    expect(screen.getByTestId('footer')).toBeInTheDocument();
  });

  it('Footer에 children이 정상적으로 렌더링된다.', () => {
    render(
      <Footer>
        <span>hi</span>
      </Footer>,
    );

    expect(screen.getByText('hi')).toBeInTheDocument();
  });

  it('Footer에 isSticky이 true일 때, position이 sticky로 설정된다.', () => {
    render(<Footer isSticky />);

    expect(screen.getByTestId('footer')).toHaveStyleRule('position', 'sticky');
  });

  it('Footer 스냅샷 테스트', () => {
    render(<Footer />);

    expect(screen.getByTestId('footer')).toMatchSnapshot();
  });
});
