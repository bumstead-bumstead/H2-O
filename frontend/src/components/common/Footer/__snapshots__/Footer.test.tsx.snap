// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Footer 컴포넌트 테스트 > Footer 스냅샷 테스트 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: flex-end;
  justify-content: flex-end;
  width: 100%;
  height: 120px;
  padding-top: 20px;
  position: fixed;
  bottom: 0;
  background-color: #FFFFFFE5;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: flex-end;
  justify-content: flex-end;
  width: 1024px;
  margin: 0 auto;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 20px;
  width: 100%;
  margin-bottom: 6px;
}

.emotion-4 {
  font-family: Hyundai Sans Text KR,sans-serif;
  font-size: 14px;
  font-weight: 500;
  line-height: 22px;
  letter-spacing: -0.03em;
  color: #00397B;
  border: 1px solid #6688B0;
  background-color: transparent;
  padding: 6px 16px;
  border-radius: 30px;
}

.emotion-4:hover {
  background-color: #A2C7E74D;
}

.emotion-4:active {
  background-color: #00397B;
  color: white;
}

.emotion-4:disabled {
  color: #838588;
  background-color: transparent;
  border-color: #DADCDD;
  cursor: not-allowed;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 8px;
}

.emotion-6 {
  color: #626466;
  font-family: Hyundai Sans Text KR,sans-serif;
  font-size: 12px;
  font-weight: 400;
  line-height: 18px;
  letter-spacing: -0.03em;
}

.emotion-7 {
  color: #00397B;
  font-family: Hyundai Sans Head KR,sans-serif;
  font-size: 24px;
  font-weight: 500;
  line-height: 32px;
  letter-spacing: -0.03em;
}

.emotion-8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-family: Hyundai Sans Text KR,sans-serif;
  font-size: 16px;
  font-weight: 500;
  line-height: 18px;
  letter-spacing: -0.03em;
  background-color: #00224A;
  color: #FFFFFF;
  border-radius: 4px;
  font-family: Hyundai Sans Head KR,sans-serif;
  font-size: 16px;
  font-weight: 500;
  line-height: 24px;
  letter-spacing: -0.03em;
  width: 340px;
  height: 44px;
}

.emotion-8:hover {
  background-color: #00397B;
}

.emotion-8:active {
  background-color: #001731;
}

.emotion-8:disabled {
  background-color: #C8CACC;
  cursor: not-allowed;
}

<footer
  class="emotion-0"
  data-testid="footer"
  height="120"
  width="100%"
>
  <div
    class="emotion-1"
  >
    <div
      class="emotion-2"
      width="fit-content"
    >
      <div
        class="emotion-3"
        width="100%"
      >
        <button
          class="emotion-4"
        >
          견적요약
        </button>
        <div
          class="emotion-5"
        >
          <p
            class="emotion-6"
            color="gray700"
          >
            현재 총 가격
          </p>
          <p
            class="emotion-7"
            color="primary500"
          >
            0 원
          </p>
        </div>
      </div>
      <button
        class="emotion-8"
      >
        다음
      </button>
    </div>
  </div>
</footer>
`;
